mllib/src/main/scala/org/apache/spark/ml/tree/codeGenerator.scala

```diff
@@ -0,0 +1,157 @@
+    evaluator.setParentClassLoader(getClass.getClassLoader)
+    evaluator.setImplementedInterfaces(implements)
+    evaluator.setClassName(clName)
+    evaluator.setDefaultImports(Array(
+      "org.apache.spark.mllib.linalg.Vectors",
+      "org.apache.spark.mllib.linalg.Vector"
+    ))
+    evaluator.cook(s"${clName}.java", code)
+    val clazz = evaluator.getClazz()
+    val endTime = System.nanoTime()
+    def timeMs: Double = (endTime - startTime).toDouble / 1000000
+    logDebug(s"Compiled Java code (${code.size} bytes) in $timeMs ms")
+    clazz
+  }
+
+  protected def freshName(): String = {
+    s"$prefix${curId.getAndIncrement}"
+  }
+
+
+  /**
+   * Convert the tree starting at the provided root node into a code generated
+   * series of if/else statements. If the tree is too large to fit in a single
+   * in-line method breaks it up into multiple methods.
+   * Returns a string for the current function body and a string of any additional
+   * functions.
+   */
+  def nodeToTree(root: Node, depth: Int): (String, String) = {
+    // Handle the different types of nodes
+    root match {
+      case node: InternalNode => {
+        // Handle trees that get too large to fit in a single in-line java method
+        depth match {
+          case 8 => {
+            val newFunctionName = freshName()
+            val newFunction = nodeToFunction(root, newFunctionName)
+            (s"return ${newFunctionName}();", newFunction)
+          }
+          case _ => {
+            val nodeSplit = node.split
+            val (leftSubCode, leftSubFunction) = nodeToTree(node.leftChild, depth + 1)
+            val (rightSubCode, rightSubFunction) = nodeToTree(node.rightChild, depth + 1)
+            val subCode = nodeSplit match {
+              case split: CategoricalSplit => {
+                val isLeft = split.isLeft
+                isLeft match {
+                  case true => s"""
+                              if (categories.contains(input.apply(${split.featureIndex}))) {
+                                ${leftSubCode}
+                              } else {
+                                ${rightSubCode}
+                              }"""
+                  case false => s"""
+                               if (categories.contains(input.apply(${split.featureIndex}))) {
+                                 ${leftSubCode}
+                               } else {
+                                 ${rightSubCode}
+                               }"""
+                }
+              }
+              case split: ContinuousSplit => {
+                s"""
+               if (input.apply(${split.featureIndex}) <= ${split.threshold}) {
+                 ${leftSubCode}
+                } else {
+                 ${rightSubCode}
+                }"""
+              }
+            }
+            (subCode, leftSubFunction + rightSubFunction)
+          }
+        }
+      }
+      case node: LeafNode => (s"return ${node.prediction};", "")
```

Comment:
Do you actually flat the tree scoring? Seems that between node to node, it's still recursion. We should have a code-gen for the entire tree.